### DISCUSSION TITLE
Why is type not narrowed when importing from type-guarded module?

### DISCUSSION THREAD

**Discussion by thomasaarholt** [microsoft/pyright#4815] [Q&A]
Given the two following files, why is the imported `bar` in `baz.py` determined to be of type `str | None` and not `str`? I'm assuming its to do with scope in some sense, and/or that pyright takes the type of the first occurrence of `bar`?

```python
# mod.py
def foo() -> str | None:
    return "foo"

bar = foo()
reveal_type(bar) # str | None

assert bar is not None
reveal_type(bar) # str
```
```python
# baz.py
from mod import bar

reveal_type(bar) # str | None - why not str?
```

**Comment by erictraut** [comment]
Type narrowing applies only within an execution scope — a function or the global scope of a module. It does not apply across functions or across modules. In your case, `bar` has an inferred type of `str | None`. It is locally (within `mod.py`) narrowed to `str` after the `assert` statement, but this narrowing does not apply outside of the module. When you import the symbol from another module, its type is its full (unnarrrowed) type.

If your intent is for `bar` to have the type `str`, you could declare it as such or you could make sure that all assignments to it assign only `str` values.

```python
# mod.py
def foo() -> str | None:
    return "foo"

_bar = foo()
reveal_type(_bar) # str | None

assert _bar is not None
bar = _bar
```
